crates/nu-utils/src/casing.rs

```diff
@@ -0,0 +1,55 @@
+use std::cmp::Ordering;
+use unicase::UniCase;
+
+pub trait IgnoreCaseExt {
+    /// Returns a [case folded] equivalent of this string, as a new String.
+    ///
+    /// Case folding is primarily based on lowercase mapping, but includes
+    /// additional changes to the source text to help make case folding
+    /// language-invariant and consistent. Case folded text should be used
+    /// solely for processing and generally should not be stored or displayed.
+    ///
+    /// Note: this method might only do [`str::to_lowercase`] instead of a
+    /// full case fold, depending on how Nu is compiled. You should still
+    /// prefer using this method for generating case-insensitive strings,
+    /// though, as it expresses intent much better than `to_lowercase`.
+    ///
+    /// [case folded]: <https://unicode.org/faq/casemap_charprop.html#2>
+    fn to_folded_case(&self) -> String;
+
+    /// Checks that two strings are a case-insensitive match.
+    ///
+    /// Essentially `to_folded_case(a) == to_folded_case(b)`, but without
+    /// allocating and copying string temporaries. Because case folding involves
+    /// Unicode table lookups, it can sometimes be more efficient to use
+    /// `to_folded_case` to case fold once and then compare those strings.
+    fn eq_ignore_case(&self, other: &str) -> bool;
+
+    /// Compares two strings case-insensitively.
+    ///
+    /// Essentially `to_folded_case(a) == to_folded_case(b)`, but without
+    /// allocating and copying string temporaries. Because case folding involves
+    /// Unicode table lookups, it can sometimes be more efficient to use
+    /// `to_folded_case` to case fold once and then compare those strings.
+    ///
+    /// Note that this *only* ignores case, comparing the folded strings without
+    /// any other collation data or locale, so the sort order may be surprising
+    /// outside of ASCII characters.
+    fn cmp_ignore_case(&self, other: &str) -> Ordering;
+}
+
+impl IgnoreCaseExt for str {
+    fn to_folded_case(&self) -> String {
+        // we only do to_lowercase, as unicase doesn't expose its case fold yet
+        // (seanmonstar/unicase#61) and we don't want to pull in another table
+        self.to_lowercase()
+    }
+
+    fn eq_ignore_case(&self, other: &str) -> bool {
+        UniCase::new(self) == UniCase::new(other)
+    }
```

Comment:
Would it make sense to have a newtype for which we implement `IgnoreCaseExt` so we can convert/fold our needles once before going through the haystacks? Else we can always go through the `to_folded_case` and the standard comparisons but that allocates for both the needle and each straw in the haystack.

Reply:
The main reason I didn't do so is that most of those cases are potentially using `contains` or `starts_with` or otherwise using operations which expect to have `&str` access (such as for natural sort). Plus, as discussed on the trait docs, for repeated comparisons it can often be preferable to fold once up front, since the folding process isn't exactly free either. ICU4X doesn't even offer case-folded equality comparison. (It does offer one-off collation, but that might still allocate internally to create sort keys.)

In short, the extra ceremony required to thread through a newtype isn't really worth it. A `Cow<str>` version which doesn't allocate if the string is already case folded might be beneficial, but no case folding provider exposes that possibility currently.

Reply:
OK makes sense, thanks for the detailed explanation.

Let's go with the definite correctness improvements this all already provides. (If profiling tells us anything egregious we can still ponder particular implementations in the future anyways)